# Med-NLP Capstone Project Modelling

### This notebook winds up EDA and starts exploratory modelling.

In [1]:
# Importing libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Reading the admissions csv
df_adm = pd.read_csv('ADMISSIONS.csv')

## Finalizing EDA:

In [2]:
df_adm.head()

ROW_ID  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME  \
0      21          22   165315  2196-04-09 12:26:00  2196-04-10 15:54:00   
1      22          23   152223  2153-09-03 07:15:00  2153-09-08 19:10:00   
2      23          23   124321  2157-10-18 19:34:00  2157-10-25 14:00:00   
3      24          24   161859  2139-06-06 16:14:00  2139-06-09 12:48:00   
4      25          25   129635  2160-11-02 02:06:00  2160-11-05 14:55:00   

  DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION  \
0       NaN      EMERGENCY       EMERGENCY ROOM ADMIT   
1       NaN       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
2       NaN      EMERGENCY  TRANSFER FROM HOSP/EXTRAM   
3       NaN      EMERGENCY  TRANSFER FROM HOSP/EXTRAM   
4       NaN      EMERGENCY       EMERGENCY ROOM ADMIT   

          DISCHARGE_LOCATION INSURANCE LANGUAGE           RELIGION  \
0  DISC-TRAN CANCER/CHLDRN H   Private      NaN       UNOBTAINABLE   
1           HOME HEALTH CARE  Medicare      NaN           CATHOLIC   
2           HOME HEALTH CARE  Medicare     ENGL           CATHOLIC   
3                       HOME   Private      NaN  PROTESTANT QUAKER   
4                       HOME   Private      NaN       UNOBTAINABLE   

  MARITAL_STATUS ETHNICITY            EDREGTIME            EDOUTTIME  \
0        MARRIED     WHITE  2196-04-09 10:06:00  2196-04-09 13:24:00   
1        MARRIED     WHITE                  NaN                  NaN   
2        MARRIED     WHITE                  NaN                  NaN   
3         SINGLE     WHITE                  NaN                  NaN   
4        MARRIED     WHITE  2160-11-02 01:01:00  2160-11-02 04:27:00   

                                           DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  \
0                            BENZODIAZEPINE OVERDOSE                     0   
1  CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS...                     0   
2                                         BRAIN MASS                     0   
3                     INTERIOR MYOCARDIAL INFARCTION                     0   
4                            ACUTE CORONARY SYNDROME                     0   

   HAS_CHARTEVENTS_DATA  
0                     1  
1                     1  
2                     1  
3                     1  
4                     1

In [3]:
df_adm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58976 entries, 0 to 58975
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   ROW_ID                58976 non-null  int64 
 1   SUBJECT_ID            58976 non-null  int64 
 2   HADM_ID               58976 non-null  int64 
 3   ADMITTIME             58976 non-null  object
 4   DISCHTIME             58976 non-null  object
 5   DEATHTIME             5854 non-null   object
 6   ADMISSION_TYPE        58976 non-null  object
 7   ADMISSION_LOCATION    58976 non-null  object
 8   DISCHARGE_LOCATION    58976 non-null  object
 9   INSURANCE             58976 non-null  object
 10  LANGUAGE              33644 non-null  object
 11  RELIGION              58518 non-null  object
 12  MARITAL_STATUS        48848 non-null  object
 13  ETHNICITY             58976 non-null  object
 14  EDREGTIME             30877 non-null  object
 15  EDOUTTIME             30877 non-null

In [4]:
df_adm.describe()

ROW_ID    SUBJECT_ID        HADM_ID  HOSPITAL_EXPIRE_FLAG  \
count  58976.000000  58976.000000   58976.000000          58976.000000   
mean   29488.500000  33755.583288  149970.809584              0.099261   
std    17025.049075  28092.726225   28883.095213              0.299014   
min        1.000000      2.000000  100001.000000              0.000000   
25%    14744.750000  11993.750000  124952.750000              0.000000   
50%    29488.500000  24133.500000  149989.500000              0.000000   
75%    44232.250000  53851.500000  174966.500000              0.000000   
max    58976.000000  99999.000000  199999.000000              1.000000   

       HAS_CHARTEVENTS_DATA  
count          58976.000000  
mean               0.973006  
std                0.162067  
min                0.000000  
25%                1.000000  
50%                1.000000  
75%                1.000000  
max                1.000000

In [5]:
df_adm.DEATHTIME.isna().sum()

53122

In [6]:
# Getting total number of deaths

no_deaths = len(df_adm)-df_adm.DEATHTIME.isna().sum()
print('Total Hospital Deaths = ' + str(no_deaths))

Total Hospital Deaths = 5854


In [7]:
df_adm.ADMISSION_TYPE.value_counts()

EMERGENCY    42071
NEWBORN       7863
ELECTIVE      7706
URGENT        1336
Name: ADMISSION_TYPE, dtype: int64

In [8]:
df_adm.DISCHARGE_LOCATION.value_counts()

HOME                         18962
HOME HEALTH CARE             13963
SNF                           7705
REHAB/DISTINCT PART HOSP      6429
DEAD/EXPIRED                  5854
LONG TERM CARE HOSPITAL       2305
SHORT TERM HOSPITAL           1534
DISC-TRAN CANCER/CHLDRN H      669
DISCH-TRAN TO PSYCH HOSP       446
HOSPICE-HOME                   402
LEFT AGAINST MEDICAL ADVI      365
HOSPICE-MEDICAL FACILITY       153
HOME WITH HOME IV PROVIDR       67
OTHER FACILITY                  63
ICF                             47
DISC-TRAN TO FEDERAL HC         11
SNF-MEDICAID ONLY CERTIF         1
Name: DISCHARGE_LOCATION, dtype: int64

In [9]:
# Getting percentage of patients except 'HOME' categorized. This means those patients are still on treatment.

pct_notokay = ((58975-18962)/58975)*100
print('Percentage of patients not recovered = ' + '%.2f' % pct_notokay + '%')

Percentage of patients not recovered = 67.85%


> __So the scope of this project would benefit around 67.85% of the hospital patients. That's a huge chunk of patients!__

In [10]:
# Converting dates

df_adm.ADMITTIME = pd.to_datetime(df_adm.ADMITTIME, format = '%Y-%m-%d %H:%M:%S', errors = 'coerce')
df_adm.DISCHTIME = pd.to_datetime(df_adm.DISCHTIME, format = '%Y-%m-%d %H:%M:%S', errors = 'coerce')
df_adm.DEATHTIME = pd.to_datetime(df_adm.DEATHTIME, format = '%Y-%m-%d %H:%M:%S', errors = 'coerce')

In [11]:
# Sorting by subject_ID and admission date

df_adm = df_adm.sort_values(['SUBJECT_ID','ADMITTIME'])
df_adm = df_adm.reset_index(drop = True)
df_adm

ROW_ID  SUBJECT_ID  HADM_ID           ADMITTIME           DISCHTIME  \
0           1           2   163353 2138-07-17 19:04:00 2138-07-21 15:48:00   
1           2           3   145834 2101-10-20 19:08:00 2101-10-31 13:58:00   
2           3           4   185777 2191-03-16 00:28:00 2191-03-23 18:41:00   
3           4           5   178980 2103-02-02 04:31:00 2103-02-04 12:15:00   
4           5           6   107064 2175-05-30 07:15:00 2175-06-15 16:00:00   
...       ...         ...      ...                 ...                 ...   
58971   58972       99985   176670 2181-01-27 02:47:00 2181-02-12 17:05:00   
58972   58973       99991   151118 2184-12-24 08:30:00 2185-01-05 12:15:00   
58973   58974       99992   197084 2144-07-25 18:03:00 2144-07-28 17:56:00   
58974   58975       99995   137810 2147-02-08 08:00:00 2147-02-11 13:15:00   
58975   58976       99999   113369 2117-12-30 07:15:00 2118-01-04 16:30:00   

      DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION  \
0           NaT        NEWBORN  PHYS REFERRAL/NORMAL DELI   
1           NaT      EMERGENCY       EMERGENCY ROOM ADMIT   
2           NaT      EMERGENCY       EMERGENCY ROOM ADMIT   
3           NaT        NEWBORN  PHYS REFERRAL/NORMAL DELI   
4           NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
...         ...            ...                        ...   
58971       NaT      EMERGENCY       EMERGENCY ROOM ADMIT   
58972       NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
58973       NaT      EMERGENCY  CLINIC REFERRAL/PREMATURE   
58974       NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
58975       NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   

              DISCHARGE_LOCATION INSURANCE LANGUAGE           RELIGION  \
0                           HOME   Private      NaN      NOT SPECIFIED   
1                            SNF  Medicare      NaN           CATHOLIC   
2      HOME WITH HOME IV PROVIDR   Private      NaN  PROTESTANT QUAKER   
3                           HOME   Private      NaN           BUDDHIST   
4               HOME HEALTH CARE  Medicare     ENGL      NOT SPECIFIED   
...                          ...       ...      ...                ...   
58971           HOME HEALTH CARE   Private     ENGL             JEWISH   
58972                       HOME   Private     ENGL      NOT SPECIFIED   
58973                        SNF  Medicare     ENGL           CATHOLIC   
58974                       HOME  Medicare     ENGL      NOT SPECIFIED   
58975                        SNF  Medicare     SPAN  JEHOVAH'S WITNESS   

      MARITAL_STATUS           ETHNICITY            EDREGTIME  \
0                NaN               ASIAN                  NaN   
1            MARRIED               WHITE  2101-10-20 17:09:00   
2             SINGLE               WHITE  2191-03-15 13:10:00   
3                NaN               ASIAN                  NaN   
4            MARRIED               WHITE                  NaN   
...              ...                 ...                  ...   
58971        MARRIED               WHITE  2181-01-26 23:35:00   
58972        MARRIED               WHITE                  NaN   
58973        WIDOWED               WHITE  2144-07-25 13:40:00   
58974        MARRIED               WHITE                  NaN   
58975      SEPARATED  HISPANIC OR LATINO                  NaN   

                 EDOUTTIME                            DIAGNOSIS  \
0                      NaN                              NEWBORN   
1      2101-10-20 19:24:00                          HYPOTENSION   
2      2191-03-16 01:10:00  FEVER,DEHYDRATION,FAILURE TO THRIVE   
3                      NaN                              NEWBORN   
4                      NaN            CHRONIC RENAL FAILURE/SDA   
...                    ...                                  ...   
58971  2181-01-27 04:18:00                                FEVER   
58972                  NaN                   DIVERTICULITIS/SDA   
58973  2144-07-25 18:50:00           RETROPERITONEAL HEMORRHAGE 

In [12]:
# Adding the next admission date and type for each subject using groupby (we have to use groupby otherwise the dates will be from different subjects)

df_adm['NEXT_ADMITTIME'] = df_adm.groupby('SUBJECT_ID').ADMITTIME.shift(-1)

# Getting the next admission type

df_adm['NEXT_ADMISSION_TYPE'] = df_adm.groupby('SUBJECT_ID').ADMISSION_TYPE.shift(-1)
df_adm

ROW_ID  SUBJECT_ID  HADM_ID           ADMITTIME           DISCHTIME  \
0           1           2   163353 2138-07-17 19:04:00 2138-07-21 15:48:00   
1           2           3   145834 2101-10-20 19:08:00 2101-10-31 13:58:00   
2           3           4   185777 2191-03-16 00:28:00 2191-03-23 18:41:00   
3           4           5   178980 2103-02-02 04:31:00 2103-02-04 12:15:00   
4           5           6   107064 2175-05-30 07:15:00 2175-06-15 16:00:00   
...       ...         ...      ...                 ...                 ...   
58971   58972       99985   176670 2181-01-27 02:47:00 2181-02-12 17:05:00   
58972   58973       99991   151118 2184-12-24 08:30:00 2185-01-05 12:15:00   
58973   58974       99992   197084 2144-07-25 18:03:00 2144-07-28 17:56:00   
58974   58975       99995   137810 2147-02-08 08:00:00 2147-02-11 13:15:00   
58975   58976       99999   113369 2117-12-30 07:15:00 2118-01-04 16:30:00   

      DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION  \
0           NaT        NEWBORN  PHYS REFERRAL/NORMAL DELI   
1           NaT      EMERGENCY       EMERGENCY ROOM ADMIT   
2           NaT      EMERGENCY       EMERGENCY ROOM ADMIT   
3           NaT        NEWBORN  PHYS REFERRAL/NORMAL DELI   
4           NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
...         ...            ...                        ...   
58971       NaT      EMERGENCY       EMERGENCY ROOM ADMIT   
58972       NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
58973       NaT      EMERGENCY  CLINIC REFERRAL/PREMATURE   
58974       NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
58975       NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   

              DISCHARGE_LOCATION INSURANCE  ...           RELIGION  \
0                           HOME   Private  ...      NOT SPECIFIED   
1                            SNF  Medicare  ...           CATHOLIC   
2      HOME WITH HOME IV PROVIDR   Private  ...  PROTESTANT QUAKER   
3                           HOME   Private  ...           BUDDHIST   
4               HOME HEALTH CARE  Medicare  ...      NOT SPECIFIED   
...                          ...       ...  ...                ...   
58971           HOME HEALTH CARE   Private  ...             JEWISH   
58972                       HOME   Private  ...      NOT SPECIFIED   
58973                        SNF  Medicare  ...           CATHOLIC   
58974                       HOME  Medicare  ...      NOT SPECIFIED   
58975                        SNF  Medicare  ...  JEHOVAH'S WITNESS   

      MARITAL_STATUS           ETHNICITY            EDREGTIME  \
0                NaN               ASIAN                  NaN   
1            MARRIED               WHITE  2101-10-20 17:09:00   
2             SINGLE               WHITE  2191-03-15 13:10:00   
3                NaN               ASIAN                  NaN   
4            MARRIED               WHITE                  NaN   
...              ...                 ...                  ...   
58971        MARRIED               WHITE  2181-01-26 23:35:00   
58972        MARRIED               WHITE                  NaN   
58973        WIDOWED               WHITE  2144-07-25 13:40:00   
58974        MARRIED               WHITE                  NaN   
58975      SEPARATED  HISPANIC OR LATINO                  NaN   

                 EDOUTTIME                            DIAGNOSIS  \
0                      NaN                              NEWBORN   
1      2101-10-20 19:24:00                          HYPOTENSION   
2      2191-03-16 01:10:00  FEVER,DEHYDRATION,FAILURE TO THRIVE   
3                      NaN                              NEWBORN   
4                      NaN            CHRONIC RENAL FAILURE/SDA   
...                    ...                                  ...   
58971  2181-01-27 04:18:00                                FEVER   
58972                  NaN                   DIVERTICULITIS/SDA   
58973  2144-07-25 18:50:00           RETROPERITONEAL HEMORRHAGE   
58974                  NaN        ABDOMINAL A

In [13]:
# Getting rows where next admission is elective and then replace with naT or nan

rows = df_adm.NEXT_ADMISSION_TYPE == 'ELECTIVE'
df_adm.loc[rows,'NEXT_ADMITTIME'] = pd.NaT
df_adm.loc[rows,'NEXT_ADMISSION_TYPE'] = np.NaN
df_adm

ROW_ID  SUBJECT_ID  HADM_ID           ADMITTIME           DISCHTIME  \
0           1           2   163353 2138-07-17 19:04:00 2138-07-21 15:48:00   
1           2           3   145834 2101-10-20 19:08:00 2101-10-31 13:58:00   
2           3           4   185777 2191-03-16 00:28:00 2191-03-23 18:41:00   
3           4           5   178980 2103-02-02 04:31:00 2103-02-04 12:15:00   
4           5           6   107064 2175-05-30 07:15:00 2175-06-15 16:00:00   
...       ...         ...      ...                 ...                 ...   
58971   58972       99985   176670 2181-01-27 02:47:00 2181-02-12 17:05:00   
58972   58973       99991   151118 2184-12-24 08:30:00 2185-01-05 12:15:00   
58973   58974       99992   197084 2144-07-25 18:03:00 2144-07-28 17:56:00   
58974   58975       99995   137810 2147-02-08 08:00:00 2147-02-11 13:15:00   
58975   58976       99999   113369 2117-12-30 07:15:00 2118-01-04 16:30:00   

      DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION  \
0           NaT        NEWBORN  PHYS REFERRAL/NORMAL DELI   
1           NaT      EMERGENCY       EMERGENCY ROOM ADMIT   
2           NaT      EMERGENCY       EMERGENCY ROOM ADMIT   
3           NaT        NEWBORN  PHYS REFERRAL/NORMAL DELI   
4           NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
...         ...            ...                        ...   
58971       NaT      EMERGENCY       EMERGENCY ROOM ADMIT   
58972       NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
58973       NaT      EMERGENCY  CLINIC REFERRAL/PREMATURE   
58974       NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
58975       NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   

              DISCHARGE_LOCATION INSURANCE  ...           RELIGION  \
0                           HOME   Private  ...      NOT SPECIFIED   
1                            SNF  Medicare  ...           CATHOLIC   
2      HOME WITH HOME IV PROVIDR   Private  ...  PROTESTANT QUAKER   
3                           HOME   Private  ...           BUDDHIST   
4               HOME HEALTH CARE  Medicare  ...      NOT SPECIFIED   
...                          ...       ...  ...                ...   
58971           HOME HEALTH CARE   Private  ...             JEWISH   
58972                       HOME   Private  ...      NOT SPECIFIED   
58973                        SNF  Medicare  ...           CATHOLIC   
58974                       HOME  Medicare  ...      NOT SPECIFIED   
58975                        SNF  Medicare  ...  JEHOVAH'S WITNESS   

      MARITAL_STATUS           ETHNICITY            EDREGTIME  \
0                NaN               ASIAN                  NaN   
1            MARRIED               WHITE  2101-10-20 17:09:00   
2             SINGLE               WHITE  2191-03-15 13:10:00   
3                NaN               ASIAN                  NaN   
4            MARRIED               WHITE                  NaN   
...              ...                 ...                  ...   
58971        MARRIED               WHITE  2181-01-26 23:35:00   
58972        MARRIED               WHITE                  NaN   
58973        WIDOWED               WHITE  2144-07-25 13:40:00   
58974        MARRIED               WHITE                  NaN   
58975      SEPARATED  HISPANIC OR LATINO                  NaN   

                 EDOUTTIME                            DIAGNOSIS  \
0                      NaN                              NEWBORN   
1      2101-10-20 19:24:00                          HYPOTENSION   
2      2191-03-16 01:10:00  FEVER,DEHYDRATION,FAILURE TO THRIVE   
3                      NaN                              NEWBORN   
4                      NaN            CHRONIC RENAL FAILURE/SDA   
...                    ...                                  ...   
58971  2181-01-27 04:18:00                                FEVER   
58972                  NaN                   DIVERTICULITIS/SDA   
58973  2144-07-25 18:50:00           RETROPERITONEAL HEMORRHAGE   
58974                  NaN        ABDOMINAL A

In [14]:
# Sorting by subject_ID and admission date (it is safer to sort right before the fill in case something changed the order above)

df_adm = df_adm.sort_values(['SUBJECT_ID','ADMITTIME'])

# Back filling values

df_adm[['NEXT_ADMITTIME','NEXT_ADMISSION_TYPE']] = df_adm.groupby(['SUBJECT_ID'])[['NEXT_ADMITTIME','NEXT_ADMISSION_TYPE']].fillna(method = 'bfill')
df_adm

ROW_ID  SUBJECT_ID  HADM_ID           ADMITTIME           DISCHTIME  \
0           1           2   163353 2138-07-17 19:04:00 2138-07-21 15:48:00   
1           2           3   145834 2101-10-20 19:08:00 2101-10-31 13:58:00   
2           3           4   185777 2191-03-16 00:28:00 2191-03-23 18:41:00   
3           4           5   178980 2103-02-02 04:31:00 2103-02-04 12:15:00   
4           5           6   107064 2175-05-30 07:15:00 2175-06-15 16:00:00   
...       ...         ...      ...                 ...                 ...   
58971   58972       99985   176670 2181-01-27 02:47:00 2181-02-12 17:05:00   
58972   58973       99991   151118 2184-12-24 08:30:00 2185-01-05 12:15:00   
58973   58974       99992   197084 2144-07-25 18:03:00 2144-07-28 17:56:00   
58974   58975       99995   137810 2147-02-08 08:00:00 2147-02-11 13:15:00   
58975   58976       99999   113369 2117-12-30 07:15:00 2118-01-04 16:30:00   

      DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION  \
0           NaT        NEWBORN  PHYS REFERRAL/NORMAL DELI   
1           NaT      EMERGENCY       EMERGENCY ROOM ADMIT   
2           NaT      EMERGENCY       EMERGENCY ROOM ADMIT   
3           NaT        NEWBORN  PHYS REFERRAL/NORMAL DELI   
4           NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
...         ...            ...                        ...   
58971       NaT      EMERGENCY       EMERGENCY ROOM ADMIT   
58972       NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
58973       NaT      EMERGENCY  CLINIC REFERRAL/PREMATURE   
58974       NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   
58975       NaT       ELECTIVE  PHYS REFERRAL/NORMAL DELI   

              DISCHARGE_LOCATION INSURANCE  ...           RELIGION  \
0                           HOME   Private  ...      NOT SPECIFIED   
1                            SNF  Medicare  ...           CATHOLIC   
2      HOME WITH HOME IV PROVIDR   Private  ...  PROTESTANT QUAKER   
3                           HOME   Private  ...           BUDDHIST   
4               HOME HEALTH CARE  Medicare  ...      NOT SPECIFIED   
...                          ...       ...  ...                ...   
58971           HOME HEALTH CARE   Private  ...             JEWISH   
58972                       HOME   Private  ...      NOT SPECIFIED   
58973                        SNF  Medicare  ...           CATHOLIC   
58974                       HOME  Medicare  ...      NOT SPECIFIED   
58975                        SNF  Medicare  ...  JEHOVAH'S WITNESS   

      MARITAL_STATUS           ETHNICITY            EDREGTIME  \
0                NaN               ASIAN                  NaN   
1            MARRIED               WHITE  2101-10-20 17:09:00   
2             SINGLE               WHITE  2191-03-15 13:10:00   
3                NaN               ASIAN                  NaN   
4            MARRIED               WHITE                  NaN   
...              ...                 ...                  ...   
58971        MARRIED               WHITE  2181-01-26 23:35:00   
58972        MARRIED               WHITE                  NaN   
58973        WIDOWED               WHITE  2144-07-25 13:40:00   
58974        MARRIED               WHITE                  NaN   
58975      SEPARATED  HISPANIC OR LATINO                  NaN   

                 EDOUTTIME                            DIAGNOSIS  \
0                      NaN                              NEWBORN   
1      2101-10-20 19:24:00                          HYPOTENSION   
2      2191-03-16 01:10:00  FEVER,DEHYDRATION,FAILURE TO THRIVE   
3                      NaN                              NEWBORN   
4                      NaN            CHRONIC RENAL FAILURE/SDA   
...                    ...                                  ...   
58971  2181-01-27 04:18:00                                FEVER   
58972                  NaN                   DIVERTICULITIS/SDA   
58973  2144-07-25 18:50:00           RETROPERITONEAL HEMORRHAGE   
58974                  NaN        ABDOMINAL A

In [15]:
df_adm['DAYS_NEXT_ADMIT']=  (df_adm.NEXT_ADMITTIME - df_adm.DISCHTIME).dt.total_seconds()/(24*60*60)

In [16]:
df_adm['DAYS_NEXT_ADMIT'].notna().sum()

11399

In [17]:
df2 = df_adm[df_adm['ADMISSION_TYPE'] == 'EMERGENCY']
df2['DAYS_NEXT_ADMIT'].notna().sum()

9545

In [18]:
# Getting percentage of unplanned readmissions of the total readmissions

pct_readmissions = (9545/11399)*100
print('Out of all the readmitted patients, a staggering ' + '%.2f' % pct_readmissions + '% were not planned!')

Out of all the readmitted patients, a staggering 83.74% were not planned!


> __So we have 11,399 hospital readmissions! Out of which 9,545 were readmitted without any prior planning! Out of all the readmitted patients, a staggering 83.74% were not planned!__

## Exploratory Modelling:

In [19]:
# Reading NOTEEVENTS csv

df_notes = pd.read_csv("NOTEEVENTS.csv")

C:\Users\Suhet\AppData\Local\Temp\ipykernel_4504\2441151186.py:3: DtypeWarning: Columns (4,5) have mixed types. Specify dtype option on import or set low_memory=False.
  df_notes = pd.read_csv("NOTEEVENTS.csv")


In [20]:
df_notes.head()

ROW_ID  SUBJECT_ID   HADM_ID   CHARTDATE CHARTTIME STORETIME  \
0     174       22532  167853.0  2151-08-04       NaN       NaN   
1     175       13702  107527.0  2118-06-14       NaN       NaN   
2     176       13702  167118.0  2119-05-25       NaN       NaN   
3     177       13702  196489.0  2124-08-18       NaN       NaN   
4     178       26880  135453.0  2162-03-25       NaN       NaN   

            CATEGORY DESCRIPTION  CGID  ISERROR  \
0  Discharge summary      Report   NaN      NaN   
1  Discharge summary      Report   NaN      NaN   
2  Discharge summary      Report   NaN      NaN   
3  Discharge summary      Report   NaN      NaN   
4  Discharge summary      Report   NaN      NaN   

                                                TEXT  
0  Admission Date:  [**2151-7-16**]       Dischar...  
1  Admission Date:  [**2118-6-2**]       Discharg...  
2  Admission Date:  [**2119-5-4**]              D...  
3  Admission Date:  [**2124-7-21**]              ...  
4  Admission Date:  [**2162-3-3**]              D...

In [21]:
df_notes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2083180 entries, 0 to 2083179
Data columns (total 11 columns):
 #   Column       Dtype  
---  ------       -----  
 0   ROW_ID       int64  
 1   SUBJECT_ID   int64  
 2   HADM_ID      float64
 3   CHARTDATE    object 
 4   CHARTTIME    object 
 5   STORETIME    object 
 6   CATEGORY     object 
 7   DESCRIPTION  object 
 8   CGID         float64
 9   ISERROR      float64
 10  TEXT         object 
dtypes: float64(3), int64(2), object(6)
memory usage: 174.8+ MB


In [22]:
# Filtering df to contain only discharge summary

df_notes_dis_sum = df_notes.loc[df_notes.CATEGORY == 'Discharge summary']

In [23]:
# Creating df with last discharge summary

df_notes_dis_sum_last = (df_notes_dis_sum.groupby(['SUBJECT_ID','HADM_ID']).nth(-1)).reset_index()
assert df_notes_dis_sum_last.duplicated(['HADM_ID']).sum() == 0, 'Multiple discharge summaries per admission'
df_notes_dis_sum_last

SUBJECT_ID   HADM_ID  ROW_ID   CHARTDATE CHARTTIME STORETIME  \
0               3  145834.0   44005  2101-10-31       NaN       NaN   
1               4  185777.0    4788  2191-03-23       NaN       NaN   
2               6  107064.0   20825  2175-06-15       NaN       NaN   
3               9  150750.0   57115  2149-11-14       NaN       NaN   
4              10  184167.0   17390  2103-07-06       NaN       NaN   
...           ...       ...     ...         ...       ...       ...   
52721       99985  176670.0   51770  2181-02-12       NaN       NaN   
52722       99991  151118.0    9682  2185-01-05       NaN       NaN   
52723       99992  197084.0   41993  2144-07-28       NaN       NaN   
52724       99995  137810.0   42710  2147-02-11       NaN       NaN   
52725       99999  113369.0   52180  2118-01-04       NaN       NaN   

                CATEGORY DESCRIPTION  CGID  ISERROR  \
0      Discharge summary      Report   NaN      NaN   
1      Discharge summary      Report   NaN      NaN   
2      Discharge summary      Report   NaN      NaN   
3      Discharge summary    Addendum   NaN      NaN   
4      Discharge summary      Report   NaN      NaN   
...                  ...         ...   ...      ...   
52721  Discharge summary      Report   NaN      NaN   
52722  Discharge summary      Report   NaN      NaN   
52723  Discharge summary      Report   NaN      NaN   
52724  Discharge summary      Report   NaN      NaN   
52725  Discharge summary      Report   NaN      NaN   

                                                    TEXT  
0      Admission Date:  [**2101-10-20**]     Discharg...  
1      Admission Date:  [**2191-3-16**]     Discharge...  
2      Admission Date: [**2175-5-30**]        Dischar...  
3      Name:  [**Known lastname 10050**], [**Known fi...  
4      Admission Date:  [**2103-6-28**]       Dischar...  
...                                                  ...  
52721  Admission Date:  [**2181-1-27**]              ...  
52722  Admission Date:  [**2184-12-24**]             ...  
52723  Admission Date:  [**2144-7-25**]              ...  
52724  Admission Date:  [**2147-2-8**]              D...  
52725  Admission Date:  [**2117-12-30**]             ...  

[52726 rows x 11 columns]

In [24]:
# Merging admission and notes df

df_adm_notes = pd.merge(df_adm[['SUBJECT_ID','HADM_ID','ADMITTIME','DISCHTIME','DAYS_NEXT_ADMIT','NEXT_ADMITTIME','ADMISSION_TYPE','DEATHTIME']],
                        df_notes_dis_sum_last[['SUBJECT_ID','HADM_ID','TEXT']], 
                        on = ['SUBJECT_ID','HADM_ID'],
                        how = 'left')
assert len(df_adm) == len(df_adm_notes), 'Number of rows increased'
df_adm_notes

SUBJECT_ID  HADM_ID           ADMITTIME           DISCHTIME  \
0               2   163353 2138-07-17 19:04:00 2138-07-21 15:48:00   
1               3   145834 2101-10-20 19:08:00 2101-10-31 13:58:00   
2               4   185777 2191-03-16 00:28:00 2191-03-23 18:41:00   
3               5   178980 2103-02-02 04:31:00 2103-02-04 12:15:00   
4               6   107064 2175-05-30 07:15:00 2175-06-15 16:00:00   
...           ...      ...                 ...                 ...   
58971       99985   176670 2181-01-27 02:47:00 2181-02-12 17:05:00   
58972       99991   151118 2184-12-24 08:30:00 2185-01-05 12:15:00   
58973       99992   197084 2144-07-25 18:03:00 2144-07-28 17:56:00   
58974       99995   137810 2147-02-08 08:00:00 2147-02-11 13:15:00   
58975       99999   113369 2117-12-30 07:15:00 2118-01-04 16:30:00   

       DAYS_NEXT_ADMIT NEXT_ADMITTIME ADMISSION_TYPE DEATHTIME  \
0                  NaN            NaT        NEWBORN       NaT   
1                  NaN            NaT      EMERGENCY       NaT   
2                  NaN            NaT      EMERGENCY       NaT   
3                  NaN            NaT        NEWBORN       NaT   
4                  NaN            NaT       ELECTIVE       NaT   
...                ...            ...            ...       ...   
58971              NaN            NaT      EMERGENCY       NaT   
58972              NaN            NaT       ELECTIVE       NaT   
58973              NaN            NaT      EMERGENCY       NaT   
58974              NaN            NaT       ELECTIVE       NaT   
58975              NaN            NaT       ELECTIVE       NaT   

                                                    TEXT  
0                                                    NaN  
1      Admission Date:  [**2101-10-20**]     Discharg...  
2      Admission Date:  [**2191-3-16**]     Discharge...  
3                                                    NaN  
4      Admission Date: [**2175-5-30**]        Dischar...  
...                                                  ...  
58971  Admission Date:  [**2181-1-27**]              ...  
58972  Admission Date:  [**2184-12-24**]             ...  
58973  Admission Date:  [**2144-7-25**]              ...  
58974  Admission Date:  [**2147-2-8**]              D...  
58975  Admission Date:  [**2117-12-30**]             ...  

[58976 rows x 9 columns]

In [28]:
# Calculating percentage of missing text admissions

m_adm = (df_adm_notes.TEXT.isnull().sum() / len(df_adm_notes))*100
print('Percentage of admissions missing text = ' + '%.2f' % m_adm + '%')

Percentage of admissions missing text = 10.60%


In [29]:
# Figuring out reason for the missing admissions

df_adm_notes.groupby('ADMISSION_TYPE').apply(lambda g: g.TEXT.isnull().sum())/df_adm_notes.groupby('ADMISSION_TYPE').size()

ADMISSION_TYPE
ELECTIVE     0.048663
EMERGENCY    0.037983
NEWBORN      0.536691
URGENT       0.042665
dtype: float64

> __Most likely, these missing NEWBORN admissions have their discharge summary stored outside of the MIMIC dataset.__ Hence dropping them.

In [30]:
# Removing NEWBORN cases from our data

df_adm_notes_clean = df_adm_notes.loc[df_adm_notes.ADMISSION_TYPE != 'NEWBORN'].copy()
df_adm_notes_clean

SUBJECT_ID  HADM_ID           ADMITTIME           DISCHTIME  \
1               3   145834 2101-10-20 19:08:00 2101-10-31 13:58:00   
2               4   185777 2191-03-16 00:28:00 2191-03-23 18:41:00   
4               6   107064 2175-05-30 07:15:00 2175-06-15 16:00:00   
7               9   150750 2149-11-09 13:06:00 2149-11-14 10:15:00   
9              11   194540 2178-04-16 06:18:00 2178-05-11 19:00:00   
...           ...      ...                 ...                 ...   
58971       99985   176670 2181-01-27 02:47:00 2181-02-12 17:05:00   
58972       99991   151118 2184-12-24 08:30:00 2185-01-05 12:15:00   
58973       99992   197084 2144-07-25 18:03:00 2144-07-28 17:56:00   
58974       99995   137810 2147-02-08 08:00:00 2147-02-11 13:15:00   
58975       99999   113369 2117-12-30 07:15:00 2118-01-04 16:30:00   

       DAYS_NEXT_ADMIT NEXT_ADMITTIME ADMISSION_TYPE           DEATHTIME  \
1                  NaN            NaT      EMERGENCY                 NaT   
2                  NaN            NaT      EMERGENCY                 NaT   
4                  NaN            NaT       ELECTIVE                 NaT   
7                  NaN            NaT      EMERGENCY 2149-11-14 10:15:00   
9                  NaN            NaT      EMERGENCY                 NaT   
...                ...            ...            ...                 ...   
58971              NaN            NaT      EMERGENCY                 NaT   
58972              NaN            NaT       ELECTIVE                 NaT   
58973              NaN            NaT      EMERGENCY                 NaT   
58974              NaN            NaT       ELECTIVE                 NaT   
58975              NaN            NaT       ELECTIVE                 NaT   

                                                    TEXT  
1      Admission Date:  [**2101-10-20**]     Discharg...  
2      Admission Date:  [**2191-3-16**]     Discharge...  
4      Admission Date: [**2175-5-30**]        Dischar...  
7      Name:  [**Known lastname 10050**], [**Known fi...  
9      Admission Date:  [**2178-4-16**]              ...  
...                                                  ...  
58971  Admission Date:  [**2181-1-27**]              ...  
58972  Admission Date:  [**2184-12-24**]             ...  
58973  Admission Date:  [**2144-7-25**]              ...  
58974  Admission Date:  [**2147-2-8**]              D...  
58975  Admission Date:  [**2117-12-30**]             ...  

[51113 rows x 9 columns]

In [31]:
# Preparing labels if readmitted < 30 days

df_adm_notes_clean['OUTPUT_LABEL'] = (df_adm_notes_clean.DAYS_NEXT_ADMIT < 30).astype('int')
df_adm_notes_clean

SUBJECT_ID  HADM_ID           ADMITTIME           DISCHTIME  \
1               3   145834 2101-10-20 19:08:00 2101-10-31 13:58:00   
2               4   185777 2191-03-16 00:28:00 2191-03-23 18:41:00   
4               6   107064 2175-05-30 07:15:00 2175-06-15 16:00:00   
7               9   150750 2149-11-09 13:06:00 2149-11-14 10:15:00   
9              11   194540 2178-04-16 06:18:00 2178-05-11 19:00:00   
...           ...      ...                 ...                 ...   
58971       99985   176670 2181-01-27 02:47:00 2181-02-12 17:05:00   
58972       99991   151118 2184-12-24 08:30:00 2185-01-05 12:15:00   
58973       99992   197084 2144-07-25 18:03:00 2144-07-28 17:56:00   
58974       99995   137810 2147-02-08 08:00:00 2147-02-11 13:15:00   
58975       99999   113369 2117-12-30 07:15:00 2118-01-04 16:30:00   

       DAYS_NEXT_ADMIT NEXT_ADMITTIME ADMISSION_TYPE           DEATHTIME  \
1                  NaN            NaT      EMERGENCY                 NaT   
2                  NaN            NaT      EMERGENCY                 NaT   
4                  NaN            NaT       ELECTIVE                 NaT   
7                  NaN            NaT      EMERGENCY 2149-11-14 10:15:00   
9                  NaN            NaT      EMERGENCY                 NaT   
...                ...            ...            ...                 ...   
58971              NaN            NaT      EMERGENCY                 NaT   
58972              NaN            NaT       ELECTIVE                 NaT   
58973              NaN            NaT      EMERGENCY                 NaT   
58974              NaN            NaT       ELECTIVE                 NaT   
58975              NaN            NaT       ELECTIVE                 NaT   

                                                    TEXT  OUTPUT_LABEL  
1      Admission Date:  [**2101-10-20**]     Discharg...             0  
2      Admission Date:  [**2191-3-16**]     Discharge...             0  
4      Admission Date: [**2175-5-30**]        Dischar...             0  
7      Name:  [**Known lastname 10050**], [**Known fi...             0  
9      Admission Date:  [**2178-4-16**]              ...             0  
...                                                  ...           ...  
58971  Admission Date:  [**2181-1-27**]              ...             0  
58972  Admission Date:  [**2184-12-24**]             ...             0  
58973  Admission Date:  [**2144-7-25**]              ...             0  
58974  Admission Date:  [**2147-2-8**]              D...             0  
58975  Admission Date:  [**2117-12-30**]             ...             0  

[51113 rows x 10 columns]

In [32]:
print('Number of positive samples:', (df_adm_notes_clean.OUTPUT_LABEL == 1).sum())
print('Number of negative samples:',  (df_adm_notes_clean.OUTPUT_LABEL == 0).sum())
print('Total samples:', len(df_adm_notes_clean))

Number of positive samples: 3004
Number of negative samples: 48109
Total samples: 51113
